simul/platform/mininet.go

```diff
@@ -0,0 +1,392 @@
+// Mininet is the platform-implementation that uses the MiniNet-framework
+// set in place by Marc-Andre Luthi from EPFL. It is based on MiniNet,
+// as it uses a lot of similar routines
+
+package platform
+
+import (
+	"fmt"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"time"
+
+	"net"
+
+	"io/ioutil"
+
+	"path"
+
+	"strings"
+
+	"sort"
+
+	"github.com/BurntSushi/toml"
+	"github.com/dedis/cothority/app/lib/config"
+	"github.com/dedis/cothority/log"
+	"github.com/dedis/cothority/sda"
+)
+
+// MiniNet represents all the configuration that is necessary to run a simulation
+// on remote hosts running Mininet.
+type MiniNet struct {
+	// *** Mininet-related configuration
+	// The login on the platform
+	Login string
+	// The outside host on the platform
+	External string
+	// Directory we start - supposed to be `cothority/simul`
+	wd string
+	// Directory holding the cothority-go-file
+	cothorityDir string
+	// Directory storing the additional files
+	mininetDir string
+	// Directory for building
+	buildDir string
+	// IPs of all hosts
+	HostIPs []string
+	// Channel to communicate stopping of experiment
+	sshMininet chan string
+	// Whether the simulation is started
+	started bool
+	// RC-configuration
+	config string
+
+	// ProxyAddress : the proxy will redirect every traffic it
+	// receives to this address
+	ProxyAddress string
+	// Port number of the monitor and the proxy
+	MonitorPort int
+
+	// Simulation to be run
+	Simulation string
+	// Number of servers to be used
+	Servers int
+	// Number of machines
+	Hosts int
+	// Debugging-level: 0 is none - 5 is everything
+	Debug int
+	// Whether to show time in debugging messages
+	DebugTime bool
+	// Whether to show color debugging-messages
+	DebugColor bool
+	// The number of seconds to wait for closing the connection
+	RunWait int
+	// Delay in ms of the network connection
+	Delay int
+	// Bandwidth in Mbps of the network connection
+	Bandwidth int
+}
+
+// Configure implements the Platform-interface. It is called once to set up
+// the necessary internal variables.
+func (m *MiniNet) Configure(pc *Config) {
+	// Directory setup - would also be possible in /tmp
+	// Supposes we're in `cothority/simul`
+	m.wd, _ = os.Getwd()
+	m.cothorityDir = m.wd + "/cothority"
+	m.mininetDir = m.wd + "/platform/mininet"
+	m.buildDir = m.mininetDir + "/build"
+	m.Login = "root"
+	log.ErrFatal(m.parseServers())
+	m.External = m.HostIPs[0]
+	m.ProxyAddress = "localhost"
+	m.MonitorPort = pc.MonitorPort
+	m.Debug = pc.Debug
+	m.DebugTime = log.ShowTime()
+	m.DebugColor = log.UseColors()
+
+	m.Delay = 0
+	m.Bandwidth = 1000
+
+	// Clean the MiniNet-dir, create it and change into it
+	os.RemoveAll(m.buildDir)
+	log.ErrFatal(os.Mkdir(m.buildDir, 0700))
+	sda.WriteTomlConfig(*m, m.mininetDir+"/mininet.toml", m.mininetDir)
+
+	if m.Simulation == "" {
+		log.Fatal("No simulation defined in runconfig")
+	}
+
+	// Setting up channel
+	m.sshMininet = make(chan string)
+}
```

Comment:
Similar comment as above: https://github.com/dedis/cothority/blob/master/simul/platform/deterlab.go#L91
Here the code duplication is far less, but I still think we could do better.